### ISSUE TITLE
panic in updateMetadataLoop : runtime error: index out of range [-1]

### ISSUE THREAD

**Issue by jcsp** [twmb/franz-go#190]
franz-go v1.6.0

I've just seen this for the first time:
```
panic: runtime error: index out of range [-1]

goroutine 1678 [running]:
github.com/twmb/franz-go/pkg/kgo.(*sink).removeRecBuf(0x7f2360?, 0xc00e80b700?)
        /root/go/pkg/mod/github.com/twmb/franz-go@v1.6.0/pkg/kgo/sink.go:978 +0x1af
github.com/twmb/franz-go/pkg/kgo.(*topicPartition).migrateProductionTo(0xc2380356e0, 0xc36611c9c0)
        /root/go/pkg/mod/github.com/twmb/franz-go@v1.6.0/pkg/kgo/topics_and_partitions.go:472 +0x35
github.com/twmb/franz-go/pkg/kgo.(*Client).mergeTopicPartitions(0xc00df57800, {0x7fff75e79cd6, 0xc}, 0xc8383be810, 0xce96771630, 0x1, 0x40e725?, 0x7f0c128963e0?, 0x0?)
        /root/go/pkg/mod/github.com/twmb/franz-go@v1.6.0/pkg/kgo/metadata.go:691 +0x1053
github.com/twmb/franz-go/pkg/kgo.(*Client).updateMetadata(0xc00df57800)
        /root/go/pkg/mod/github.com/twmb/franz-go@v1.6.0/pkg/kgo/metadata.go:354 +0x7ee
github.com/twmb/franz-go/pkg/kgo.(*Client).updateMetadataLoop(0xc00df57800)
        /root/go/pkg/mod/github.com/twmb/franz-go@v1.6.0/pkg/kgo/metadata.go:184 +0x5a6
created by github.com/twmb/franz-go/pkg/kgo.NewClient
        /root/go/pkg/mod/github.com/twmb/franz-go@v1.6.0/pkg/kgo/client.go:213 +0x9bc
```

The exact code in use is https://github.com/jcsp/kgo-verifier/tree/c6540c4b68d6a81a9677fa66ebd9e4576aecc1d9

It's a traffic generator that runs a bunch of threads with their own client, a producer+consumer in each one, producing one message for each consumed.

The cluster under test is redpanda nightly 0.0.0~20220809gitfbdafab-1 , running with 9 nodes and 67490 partitions in 1 topic.

**Comment by twmb** [comment]
This is perplexing

**Comment by twmb** [comment]
I'm getting close to calling this one blocked until it's reproduced. I'm not sure how this could possibly happen.

* The corresponding panic would happen if a `recBufsIdx` field is -1: https://github.com/twmb/franz-go/blob/e4bd2620cfcad922de7eec287bcb974f488f6cac/pkg/kgo/sink.go#L978
* The only place a `recBuf` is initialized (which is where this field is initialized -1) is in metadata.go: https://github.com/twmb/franz-go/blob/e4bd2620cfcad922de7eec287bcb974f488f6cac/pkg/kgo/metadata.go#L456
* This is initialized once we receive a metadata response. We use it to update everything internally: https://github.com/twmb/franz-go/blob/e4bd2620cfcad922de7eec287bcb974f488f6cac/pkg/kgo/metadata.go#L289
* We eventually use this new load to merge it into the existing data: https://github.com/twmb/franz-go/blob/e4bd2620cfcad922de7eec287bcb974f488f6cac/pkg/kgo/metadata.go#L347
* The last thing we do in the merge is for any partition not on a sink already, we add it to the leader sink: https://github.com/twmb/franz-go/blob/e4bd2620cfcad922de7eec287bcb974f488f6cac/pkg/kgo/metadata.go#L705-L706
* This specific panic is from `migrateProductionTo`, which first removes a partition from its old sink and adds to a new sink, this happens in the metadata loop: https://github.com/twmb/franz-go/blob/e4bd2620cfcad922de7eec287bcb974f488f6cac/pkg/kgo/metadata.go#L691
* and the relevant move logic is here: https://github.com/twmb/franz-go/blob/e4bd2620cfcad922de7eec287bcb974f488f6cac/pkg/kgo/topics_and_partitions.go#L470-L493
* The first time a topic's metadata is requested, there will be no previous partitions, so _only_ the `if -1, add to sink` logic runs.
* * Adding to a sink updates the `recBufsIdx` to non-negative: https://github.com/twmb/franz-go/blob/e4bd2620cfcad922de7eec287bcb974f488f6cac/pkg/kgo/sink.go#L965
* Saving partitions for a topic happens at the end of the merge: https://github.com/twmb/franz-go/blob/e4bd2620cfcad922de7eec287bcb974f488f6cac/pkg/kgo/metadata.go#L539

Notably, moving only happens when the partition exists. The partition can only exist once it is stored for the first time. Just before the store, all new partitions (-1 `recBufsIdx`) are added to sinks, and adding to a sink ensures the index is non-negative. I'm struggling to see a path in the code such that there is an existing partition that was not added to a sink that is now migrating to a new sink. I'll stare a bit more, but this might just be blocked until we have another panic. This code also has not changed in ~2yr. I never like to blame the system, but I wonder if something particularly odd happened that caused one of the steps to be skipped (which should be impossible).

**Comment by twmb** [comment]
Setting this to blocked until we have either a reproducer or another panic (with hopefully more goroutines / logs) to continue an investigation from.

**Comment by brunsgaard** [comment]
@twmb your efforts looking into issue reports are admirable 👍

**Comment by jcsp** [comment]
This reocurred today twice when using 3191842a81033342e8d37a529bd0a1b3d190fd9f

I don't think it had anything to do with that specific commit, although when testing I was jumping all the way from 1.6.0 to 1.7.0+that fix.

This happened right at the start, after my test waits for a consumer group to be stable and then starts producing: as soon as the producer started, it hit this panic.

For completeness here's the latest stack trace:

```
panic: runtime error: index out of range [-1]

goroutine 715 [running]:
github.com/twmb/franz-go/pkg/kgo.(*sink).removeRecBuf(0xc04bfedcb0, 0x8f7328)
        /home/vectorized/go/pkg/mod/github.com/twmb/franz-go@v1.7.1-0.20220901035118-3191842a8103/pkg/kgo/sink.go:994 +0x1af
github.com/twmb/franz-go/pkg/kgo.(*topicPartition).migrateProductionTo(0xc0139aa630, 0xc04c7496e0)
        /home/vectorized/go/pkg/mod/github.com/twmb/franz-go@v1.7.1-0.20220901035118-3191842a8103/pkg/kgo/topics_and_partitions.go:472 +0x35
github.com/twmb/franz-go/pkg/kgo.(*Client).mergeTopicPartitions(0xc004fee800, {0x7ffc4fe7ecb9, 0xc}, 0xc00b882db0, 0xc046732ff0, 0x1, 0x8f8790, 0x7ce2c0, 0xc02a829e10)
        /home/vectorized/go/pkg/mod/github.com/twmb/franz-go@v1.7.1-0.20220901035118-3191842a8103/pkg/kgo/metadata.go:726 +0x1085
github.com/twmb/franz-go/pkg/kgo.(*Client).updateMetadata(0xc004fee800)
        /home/vectorized/go/pkg/mod/github.com/twmb/franz-go@v1.7.1-0.20220901035118-3191842a8103/pkg/kgo/metadata.go:390 +0x7ee
github.com/twmb/franz-go/pkg/kgo.(*Client).updateMetadataLoop(0xc004fee800)
        /home/vectorized/go/pkg/mod/github.com/twmb/franz-go@v1.7.1-0.20220901035118-3191842a8103/pkg/kgo/metadata.go:220 +0x5a6
created by github.com/twmb/franz-go/pkg/kgo.NewClient
        /home/vectorized/go/pkg/mod/github.com/twmb/franz-go@v1.7.1-0.20220901035118-3191842a8103/pkg/kgo/client.go:214 +0x985
```

**Comment by jcsp** [comment]
Unfortunately trying to reproduce it with trace logging did not hit it.

**Comment by twmb** [comment]
I saw the issue, this was ticky. A full writeup is in 0ca6478600c632deed4c7d65c13c3459d19071bd

I'll wait ~1 or 2 days to see if you find any more issues before I tag v1.7.1 (perhaps on Monday)

**Comment by twmb** [comment]
I tested this fix by manually hooking into metadata across three loads:

* first consume, causes one metadata load, this is normal
* next, produce, causes another metadata load, hook into it and set the load error to non-nil
* on the third load, change the leader

Before this commit, I saw the same panic as in this issue. After this commit, I saw no panic.

**Comment by twmb** [comment]
This fix has been released in v1.7.1
